File: Compose/YABA/app/src/main/java/dev/subfly/yaba/core/state/selection/icon/IconCategorySelectionStateMachine.kt
package dev.subfly.yaba.core.state.selection.icon

import dev.subfly.yaba.core.managers.IconManager
import dev.subfly.yaba.core.state.base.BaseStateMachine

class IconCategorySelectionStateMachine :
    BaseStateMachine<IconCategorySelectionUIState, IconCategorySelectionEvent>(
        initialState = IconCategorySelectionUIState(),
    ) {

    private var isInitialized = false

    override fun onEvent(event: IconCategorySelectionEvent) {
        when (event) {
            IconCategorySelectionEvent.OnInit -> onInit()
        }
    }

    private fun onInit() {
        if (isInitialized) return
        isInitialized = true
        launch {
            updateState { it.copy(isLoading = true) }
            val categories = IconManager.loadAllCategories()
            updateState {
                it.copy(
                    categories = categories,
                    isLoading = false,
                )
            }
        }
    }

    override fun clear() {
        isInitialized = false
        super.clear()
    }
}

File: Darwin/YABACore/State/Selection/IconCategory/IconCategorySelectionStateMachine.swift
//
//  IconCategorySelectionStateMachine.swift
//  YABACore
//

import Foundation

/// Parity with Compose `IconCategorySelectionStateMachine`.
@MainActor
public final class IconCategorySelectionStateMachine: YabaBaseObservableState<IconCategorySelectionUIState>, YabaScreenStateMachine {
    private var didLoadCategories = false

    public override init(initialState: IconCategorySelectionUIState = IconCategorySelectionUIState()) {
        super.init(initialState: initialState)
    }

    public func send(_ event: IconCategorySelectionEvent) async {
        switch event {
        case .onInit:
            guard !didLoadCategories else { return }
            didLoadCategories = true
            apply { $0.isLoading = true }
            let categories = await IconManager.loadAllCategories()
            apply {
                $0.categories = categories
                $0.isLoading = false
            }
        }
    }
}
